gems/smithy-client/lib/smithy-client/param_validator.rb

```diff
@@ -53,33 +49,40 @@ def valid_structure?(shape, values, errors, context)
       def union(shape, values, errors, context)
         return unless valid_union?(shape, values, errors, context)
 
-        values.each_pair do |name, value|
-          next if value.nil?
+        if values.is_a?(Schema::Union)
+          member_shape = shape.member_by_type(values.class)
+          shape(member_shape.shape, values.value, errors, context)
+        elsif values.respond_to?(:each_pair)
+          values.each_pair do |name, value|
+            next if value.nil?
 
-          if shape.member?(name)
-            member_shape = shape.member(name)
-            shape(member_shape.shape, value, errors, context + "[#{name.inspect}]")
-          else
-            errors << "unexpected value at #{context}[#{name.inspect}]"
+            member(shape, name, value, errors, context)
           end
         end
       end
 
       def valid_union?(shape, values, errors, context)
```

Comment:
I think we could give the user a raise when the given object is not an instance of Union's subclasses? Something like [here](https://github.com/smithy-lang/smithy-ruby/blob/main/codegen/projections/rails_json/lib/rails_json/builders.rb#L871-L872)?

Reply:
The param validator should already be doing that in valid_union?